### PR TITLE
Track sealing processes across lotus-miner restarts

### PR DESCRIPTION
**Pull request by magik6k**
Summary of changes here:
* Worker calls are now async
  * When called worker methods
    * Generate and persist a CallID (basically a random UUID)
    * Start the work in a goroutine
    * Return the CallID to manager
    * After work is finished call Return[...] with the CallID and the result on the Manager
* Manager is now tracking work and calls across restarts
  * When a method is called on the manager, it generates WorkID (tuple of task type (like PreCommit1) and params)
  * If we see that this WorkID is not running we persist it in work statestore (set work state as `started`)
  * We schedule the task as before, after the scheduler allocates us to a worker, we execute the worker call
  * We connect the CallID we got from the worker to the WorkID, and set the work state to `running`
  * After the worker returns, we translate the CallID to WorkID, and pipe the result to the correct place, mark the Work as done, and the Call as returned
* Manager <-> Worker RPCs are now using HTTP instead of websockets, not having to rely on a single TCP connection serviving for days is much more robust

### CODE REVIEW

File: extern/sector-storage/sched.go [hunk 1]
```diff
@@ -263,6 +265,9 @@ func (sh *scheduler) runSched() {
 		loop:
 			for {
 				select {
+				case <-sh.workerChange:
```
Comment (whyrusleeping) [inline extern/sector-storage/sched.go:268]: Worker changes are unhandled?

Comment (whyrusleeping) [reply extern/sector-storage/sched.go:268]: if this is 'okay' at least add a comment here explaining why

Comment (magik6k) [reply extern/sector-storage/sched.go:268]: It's a `chan struct{}`, this channel just pokes the scheduler to re-run, which is all that we need to do here

File: extern/sector-storage/sched.go [hunk 2]
```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
...
+
 		for {
-			// ask for more windows if we need them
-			for ; windowsRequested < SchedWindows; windowsRequested++ {
+			sh.workersLk.Lock()
+			enabled := worker.enabled
+			sh.workersLk.Unlock()
+
+			// ask for more windows if we need them (non-blocking)
+			for ; enabled && windowsRequested < SchedWindows; windowsRequested++ {
```
Comment (whyrusleeping) [inline extern/sector-storage/sched.go:636]: the nesting here is really painful. Can we maybe break this up? (can be in a followup PR, but should happen soon to make this logic more readable IMO)

File: extern/sector-storage/sched.go [hunk 3]
```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
...
+		closingMgr: make(chan struct{}),
+		closedMgr:  make(chan struct{}),
+	}
+
+	wid := WorkerID(sessID)
+
+	sh.workersLk.Lock()
+	_, exist := sh.workers[wid]
+	if exist {
+		// this is ok, we're already handling this worker in a different goroutine
```
Comment (whyrusleeping) [inline extern/sector-storage/sched.go:566]: is this notable at least?

File: extern/sector-storage/sched.go [hunk 4]
```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
...
-				return
-			case <-workerClosing:
-				return
-			case <-worker.closingMgr:
-				return
+			// wait for more windows to come in, or for tasks to get finished (blocking)
+			for {
+
+				// first ping the worker and check session
+				{
```
Comment (whyrusleeping) [inline extern/sector-storage/sched.go:653]: Lets break this subscope into a separate function, its really hard to tell whats being done here, and what the potential outcomes are (and why). I see several returns, a continue, and a few routes that just flow through. We should probably have a function that returns some value to signal whether to loopcontinue, return, or keep going

File: extern/sector-storage/sched.go [hunk 5]
```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
...
+
+				select {
+				case <-heartbeatTimer.C:
+					continue
+				case w := <-scheduledWindows:
+					worker.wndLk.Lock()
+					worker.activeWindows = append(worker.activeWindows, w)
+					worker.wndLk.Unlock()
+				case <-taskDone:
+					log.Debugw("task done", "workerid", wid)
```
Comment (whyrusleeping) [inline extern/sector-storage/sched.go:713]: what even is this flow control? continues returns *and* breaks?

File: extern/sector-storage/worker_local.go [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+
+	ct          *workerCallTracker
+	acceptTasks map[sealtasks.TaskType]struct{}
+	running     sync.WaitGroup
+
+	session uuid.UUID
+	closing chan struct{}
+}
+
+func newLocalWorker(executor func() (ffiwrapper.Storage, error), wcfg WorkerConfig, store stores.Store, local *stores.Local, sindex stores.SectorIndex, ret storiface.WorkerReturn, cst *statestore.StateStore) *LocalWorker {
```
Comment (whyrusleeping) [inline extern/sector-storage/worker_local.go:55]: mild preference for a named ExecutorFunc type

**Review by whyrusleeping — APPROVED** [review]
Alright, so i left a number of comments, and wasnt able to do a good job reviewing worker_local.go, but overall this LGTM.

**Review by arajasek — APPROVED** [review]
Okay, I'll call this "approved" after having spent a couple more hours on it, but that basically means "I think I understand how this is supposed to work". Ish.

File: extern/sector-storage/sched.go (hunk 3 above)
Comment (magik6k) [reply extern/sector-storage/sched.go:566]: I don't think this can happen outside of a test, potentially worth logging

**Review by Kubuxu — APPROVED** [review]
SGWM, but if there is a subtle bug here, I've for sure missed it.

**Review by whyrusleeping — APPROVED** [review]

**Comment by magik6k** [comment]
Closes #3453
Closes #3585
Closes #3597
Closes #3672
Closes #3744
Closes #3814
Closes #3815
Closes #3825
Closes #4262

**Comment by cryptowhizzard** [comment]
Seems my issue is not resolved yet ( on the latest master build ).

2020-11-10T13:42:13.729 INFO filecoin_proofs::api > generate_piece_commitment:start
2020-11-10T13:42:14.093 INFO filecoin_proofs::api > generate_piece_commitment:finish
2020-11-10T13:42:14.098 INFO filecoin_proofs::api > generate_piece_commitment:start
2020-11-10T13:42:14.173 INFO filecoin_proofs::api > generate_piece_commitment:finish
2020-11-10T13:42:14.180 INFO filecoin_proofs::api > generate_piece_commitment:start
2020-11-10T13:42:17.729Z        ^[[34mINFO^[[0m miner   miner/miner.go:382      Time delta between now and our mining base: 47s (nulls: 0)
2020-11-10T13:42:18.176Z        ^[[34mINFO^[[0m rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/client.go:346   rpc output message buffer       {"n": 4}
2020-11-10T13:42:18.367Z        ^[[34mINFO^[[0m miner   miner/miner.go:382      Time delta between now and our mining base: 48s (nulls: 1)
2020-11-10T13:42:18.419Z        ^[[34mINFO^[[0m rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/client.go:346   rpc output message buffer       {"n": 5}
2020-11-10T13:43:13.667Z        ^[[31mERROR^[[0m        rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:654        Connection timeout      {"remote": "127.0.0.1:1234"}
2020-11-10T13:43:13.667Z        ^[[31mERROR^[[0m        storageminer    storage/wdpost_run.go:751       handler: websocket connection closed
2020-11-10T13:43:13.667Z        ^[[33mWARN^[[0m storageminer    storage/wdpost_sched.go:106     window post scheduler notifs channel closed
2020-11-10T13:43:13.667Z        ^[[31mERROR^[[0m        events  events/events_called.go:382     event diff fn failed: handler: websocket connection closed
2020-11-10T13:43:13.667Z        ^[[33mWARN^[[0m events  events/events.go:97     listenHeadChanges quit
2020-11-10T13:43:13.667Z        ^[[33mWARN^[[0m rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:289        failed to send request  {"method": "xrpc.cancel", "id": 2361, "error": "write tcp 127.0.0.1:38560->127.0.0.1:1234: use of closed network connection"}
2020-11-10T13:43:13.669Z        ^[[31mERROR^[[0m        storageminer    storage/wdpost_sched.go:94      ChainNotify error: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[31mERROR^[[0m        events  events/events_called.go:478     getting parent tipset in checkNewCalls: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[33mWARN^[[0m events  events/events.go:155    headChange failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[31mERROR^[[0m        events  events/events_called.go:382     event diff fn failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[31mERROR^[[0m        events  events/events_called.go:478     getting parent tipset in checkNewCalls: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[33mWARN^[[0m events  events/events.go:155    headChange failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:13.669Z        ^[[33mWARN^[[0m events  events/events.go:97     listenHeadChanges quit
2020-11-10T13:43:13.669Z        ^[[33mWARN^[[0m rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:289        failed to send request  {"method": "xrpc.cancel", "id": 2349, "error": "write tcp 127.0.0.1:38560->127.0.0.1:1234: use of closed network connection"}
2020-11-10T13:43:14.669Z        ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-10T13:43:14.669Z        ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-10T13:43:14.669Z        ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-10T13:43:14.669Z        ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting

**Pull request merged**